File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        ListNode* tort = head;
        ListNode* hare = head;
        while(hare->next and hare->next->next){
            hare = hare->next->next;
            tort = tort->next;
            if(hare == tort) break;
        }
        tort = head;
        if(hare == nullptr) return false;
        while(hare->next and hare->next->next){        
            hare = hare->next->next;
            tort = tort->next;
            if(hare == tort) return true;
        }
        return false;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        falseHead = ListNode(0)
        falseHead.next = head
        slow, fast = falseHead, falseHead
        for i in range(n):
            fast = fast.next
        print(fast.val)
        while fast.next:
            slow = slow.next
            fast = fast.next
        tmp = slow.next
        print(tmp.val)
        slow.next = slow.next.next
        return falseHead.next